### PR TITLE
fix: prevent commentor/viewer from enabling hidden columns

### LINKED ISSUE #7867: 🐛 Bug: Commenter/Viewer can enable columns in View but they show empty
### Please confirm if bug report does NOT exist already ?

- [X] I confirm there is no existing issue for this

### Steps to reproduce ?

1. As SuperAdmin you hide some columns.
2. As Commenter/Viewer you can show these columns but the values in these columns are empty.

### Desired Behavior

Commenter/Viewer shouldn’t see hidden columns in the Fields menu. 

I can fix this bug if you agree with this proposal.

### Project Details

Node: v18.18.2
Docker: false
RootDB: Postgres
PackageVersion: 0.204.5

### Attachments

https://github.com/nocodb/nocodb/assets/162738418/5c1d35df-aa39-461b-9715-801e1ed8e0e2

### PR DESCRIPTION
**Pull request by sreehari2003**
## Change Summary
- Closes #7867 
Provide summary of changes with issue number if any.
fix: prevent commentor/viewer from enabling hidden columns
- Video https://sreehari-jayaraj.neetorecord.com/watch/399d7ab6-8a6c-42c8-9bf7-f279311e221b

## Change type

- [ ] feat: (new feature for the user, not a new feature for build script)
- [x] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of

## Summary by CodeRabbit

- **New Features**
	- Added role-based visibility for fields in the toolbar, enhancing user interface customization based on project roles.
	- Introduced conditional visibility for columns in views, allowing users to specify exceptions via an optional parameter.

### CODE REVIEW

**Review by dstala — CHANGES_REQUESTED** [review]
For viewer/commenter role, shouldn't display the fields count - misleading
Screenshot 2024-04-16 at 12 51 42 PM

**Comment by sreehari2003** [comment]
>   fix: prevent commentor/viewer from enabling hidden columns
> * Video https://sreehari-jayaraj.neetorecord.com/watch/399d7ab6-8a6c-42c8-9bf7-f279311e221b
> * This will break when the user clicks on the show all fields button , when someone clicks on the view all fields button an API call is triggered to convert all columns shown to `true`. we could hide the view and hide columns button  from read-only access users. please let me know if we that issue can be fixed by some other way.
n


@dstala  I have fixed the hidden count issue and the issue I mentioned here.
video -  https://sreehari-jayaraj.neetorecord.com/watch/dc028659-0a00-49aa-8fd6-64c7f685aa70

File: packages/nc-gui/composables/useViewColumns.ts [hunk 1]
```diff
@@ -111,12 +111,19 @@ const [useProvideViewColumns, useViewColumns] = useInjectionState(
       }
     }
 
-    const showAll = async (ignoreIds?: any) => {
+    const showAll = async (ignoreIds: Array<string> = []) => {
+      console.log(ignoreIds)
```
Comment (dstala) [inline packages/nc-gui/composables/useViewColumns.ts:115]: clean up debug log

**Review by dstala — CHANGES_REQUESTED** [review]
for viewer, the fields count is always hidden. we should exclude default count. but if user hides some more fields explicitly, count for those should be visible

**Comment by sreehari2003** [comment]
> for viewer, the fields count is always hidden. we should exclude default count. but if user hides some more fields explicitly, count for those should be visible

@dstala  I have added this feature. Please see this video - https://sreehari-jayaraj.neetorecord.com/watch/ec9fdbd5-24d8-4e2f-aa78-d6b2f7cea42e?time=62

**Review by dstala — CHANGES_REQUESTED** [review]
the solution is broken for editor & creator roles. the said changes should only be applied to viewer/commenter role.

**Comment by sreehari2003** [comment]
> the solution is broken for editor & creator roles. the said changes should only be applied to viewer/commenter role. that likely explains why tests are failing. you can investigate further.

I have fixed the issue. Right now only one test in SQLite is failing but I think it is not associated with this PR.

**Pull request closed without merging**